alt_src/alt_src.py

```diff
@@ -1971,17 +1971,33 @@ def acquire_lock(lock_file_path, wait_time, sleep_interval, logger):
 
 
 def explode_srpm(srpm, destdir=None, logfile=None):
-    # explode our srpm to the given directory
+    # explode our srpm to the given directory using the best possible means:
+    # - by installing the RPM
+    # - or falling back to rpm2cpio
+
     header = koji.get_rpm_header(srpm)
     if header[rpm.RPMTAG_SOURCEPACKAGE] != 1:
         # we checked this earlier, but since we're about to rpm -i it,
         # let's check again
         raise SanityError("%s is not a source package" % srpm)
+
     if destdir is None:
         destdir = os.getcwd()
     else:
         destdir = os.path.abspath(destdir)
         koji.ensuredir(destdir)
+
+    try:
+        return explode_srpm_install(srpm, destdir, logfile)
+    except CommandError:
+        logging.getLogger("altsrc").exception("Could not install SRPM, trying rpm2cpio")
+
+    return explode_srpm_cpio(srpm, header, destdir, logfile)
+
+
+def explode_srpm_install(srpm, destdir=None, logfile=None):
+    # explode our srpm to the given directory by installing it
+
     cmd = ['rpm', '--nosignature', '-i', '--define', '_topdir %s' % destdir, srpm]
```

Comment:
I'd suggest also fixing the "macro has empty body" problem since we are getting back to rpm -i command.
It seems that written like this it is run like:
`rpm  --nosignature -i --define _topdir /var/cache/altsrc/staging/stage/c8-stream-rhel8/rpms/r/rust/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81/checkout /mnt/redhat/brewroot/vol/rhel-8/packages/rust/1.43.1/1.module+el8.2.1+6788+cceb3c81/src/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81.src.rpm`

which produces:
`error: Macro %_topdir has empty body
error: Macro %_topdir has empty body
error: Macro %_topdir has empty body
D: ============== /var/cache/altsrc/staging/stage/c8-stream-rhel8/rpms/r/rust/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81/checkout
error: not an rpm package`
You may notice that the destdir string is taken as srpm.

So I guess the macro expression _topdir %s should be explicitly enclosed by single quotes as man rpm suggests: --define='MACRO EXPR'
then it behaves better.

`rpm --nosignature -i --define '_topdir /var/cache/altsrc/staging/stage/c8-stream-rhel8/rpms/r/rust/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81/checkout' /mnt/redhat/brewroot/vol/rhel-8/packages/rust/1.43.1/1.module+el8.2.1+6788+cceb3c81/src/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81.src.rpm
D: ============== /mnt/redhat/brewroot/vol/rhel-8/packages/rust/1.43.1/1.module+el8.2.1+6788+cceb3c81/src/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81.src.rpm
D: Expected size:     94461749 = lead(96)+sigs(4404)+pad(4)+data(94457245)
D:   Actual size:     94461749
D: /mnt/redhat/brewroot/vol/rhel-8/packages/rust/1.43.1/1.module+el8.2.1+6788+cceb3c81/src/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81.src.rpm: Header SHA1 digest: OK (00b1c5f6b6db49d910fc6d061affc02bfb68f87d)
D: 	added source package [0]
D: found 1 source and 0 binary packages`

I'm not sure what's the root cause of this (maybe length of string?)

Reply:
Did you see that "empty body error" from alt-src, or only when running RPM directly?  I saw what you posted on the issue, but I kinda suspected it was a mistake caused by using the shell to simulate what alt-src would run.

Note the command is already defined as:

```
['rpm', '--nosignature', '-i', '--define', '_topdir %s' % destdir, srpm]
```

See that '_topdir <destdir>' is already one argument which contains whitespace.  That's expected to be passed through to the subprocess correctly, and we're not using a shell here.  If you literally wrote this into a shell:

```
rpm --nosignature -i --define _topdir /var/cache/altsrc/staging/stage/c8-stream-rhel8/rpms/r/rust/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81/checkout /mnt/redhat/brewroot/vol/rhel-8/packages/rust/1.43.1/1.module+el8.2.1+6788+cceb3c81/src/rust-1.43.1-1.module+el8.2.1+6788+cceb3c81.src.rpm
```

...then what you were running isn't the same as what alt-src would have run.  You would have to use quotes like '_topdir ...', but that's a shell issue which doesn't affect usage of subprocess.Popen (if the command is prepared as a list rather than a single string, and if shell=True is not used).

If we did put quotes into the argument, like this:

```
['rpm', '--nosignature', '-i', '--define', "'_topdir %s'" % destdir, srpm]
```

...I'm pretty sure that would break the code, not fix it.

Could you have a look and review whether you still think this needs to be changed?

Reply:
Because there is not much info about the error in Pub log, I was under impression, that single quotes might be the issue that hid the actual problem. But I checked alt-src logs, they include more info and they don't complain with macro problem but rather with actual issue with rpmlib.
Also checked manually with short script on pub worker, it works as expected as is.